hnn_core/opt_toy_example/general.py

```diff
@@ -0,0 +1,378 @@
+import numpy as np
+from hnn_core import simulate_dipole
+from hnn_core.network import pick_connection
+from skopt import gp_minimize
+from skopt.utils import use_named_args
+from skopt.space import Real
+from scipy.optimize import fmin_cobyla
+
+
+class Optimizer:
+    def __init__(self, net, constraints, solver, metric):
+        self.net = net
+        self.constraints = constraints
+        if solver == 'bayesian':
+            self._get_params = _get_params_bayesian
+            self._run_opt = _run_opt_bayesian
+        elif solver == 'cobyla':
+            self._get_params = _get_params_cobyla
+            self._run_opt = _run_opt_cobyla
+        if metric == 'evoked':
+            self.metric = _rmse_evoked
+        elif metric == 'rhythmic':
+            self.metric = _rmse_rhythmic
+        self.net_ = None
+        self.error_values = None
+        self.opt_params = None
+
+    def fit(self, target_statistic, window_len=None):
+        """ ...
+            Parameters
+            ----------
+            target_statistic : dpl
+            window_len : for smoothing dpl
+            """
+
+        init_params, cons = self._get_params(self.net, self.constraints)
+        opt_params, error_values = self._run_opt(self.net,
+                                                 init_params,
+                                                 cons,
+                                                 self.metric,
+                                                 target_statistic,
+                                                 window_len)
+        self.opt_params = opt_params
+        self.error_values = error_values
+        # self.net_ = _set_params(opt_params)
+        return
+
+    def plot_convergence():
+        return
+
+    def plot_param_search():
+        return
+
+
+# gets called only once
+def _get_params_bayesian(net, constraints):
```

Comment:
it might be cleaner to do:

```py

def _get_params(net, params):
     # insert logic of flattening network
     # into a dictionary
    return params, constraints

def _get_params_bayesian(net, params):
    params, constraints = _get_params(net, constraints)
    for param in params:
        # now convert into solver-specific format
        continue

def _get_params_cobyla(net, params):
    params, constraints = _get_params(net, constraints)
    for param in params:
        # now convert into solver-specific format
        continue
```

so you don't repeat the logic of flattening multiple times and introduce potential sources of error